topics/threaded-io.md

```diff
@@ -0,0 +1,127 @@
+                                                    +-----------------+
+                                                    | ...             |
+                                                    +-----------------+
+
+
+```
+
+### Dispatch read/write tasks to I/O threads
+So when the `server.clients_pending_read` is ready (same for the writing process), we need to dispatch them to the I/O threads which are waiting.
+
+Redis prepared a global variable `io_threads_list` to store the task. `io_threads_list` is a list type variable with thread number length (configured by `io-threads n`). Its item value is a list as well which each sub-item stores the `client` object pointer.
+
+`io_threads_list[i]` represents the task list (client list) the i-th thread should handle. Redis dispatch the pending client using **Round-Robin** algorithm. So when we configured **4 I/O threads** (thread 0 to thread 3), and have **7 clients** (client 0 to client6) in `server.clients_pending_read` list, they will be dispatched like:
+```
+io_threads_list
++---------------+------------+------------+------------+
+|   thread 0    |  thread 1  |  thread 2  |  thread 3  |
+| (main thread) |(I/O thread)|(I/O thread)|(I/O thread)|
++---------------+------------+------------+------------+
+      \/              \/           \/            \/        
+      \/              \/           \/            \/             
+      \/              \/           \/            \/        
+  +----------+   +----------+  +----------+  +----------+
+  | client 0 |   | client 1 |  | client 2 |  | client 3 |
+  +----------+   +----------+  +----------+  +----------+
+  | client 4 |   | client 5 |  | client 6 |
+  +----------+   +----------+  +----------+
+
+```
+
+Each thread can iterate over the clients they should read, do read action, and store the result to `client.querybuf` attribute. The main thread will also handle part of the pending tasks after dispatching. In this example, thread 0, which is the main thread, will read through client 0 and then client 4, thread 1 will handle client 1 and client 5, and so on.
+
+### Blocking and Locking
+There will be a racing condition if the I/O thread keeps monitoring and editing the `io_threads_list[i]` list. So Redis prepared another global variable named `io_threads_pending`. It's a thread number length list and stores an integer value representing the number of the remaining task that thread i should handle. When the main thread finished dispatching task to `io_threads_list`, it set `io_threads_pending[i]` to the actual task count.
+
+The I/O threads will stay in a `for (int j = 0; j < 1000000; j++)` loop and checking if `io_threads_pending[id] != 0`. So when the value changed to an positive integer, thread i begin to read/write through the client in `io_threads_list[i]`. Once I/O threads finish their job, they will empty the list in `io_threads_list[i]`, and set `io_threads_pending[i]` to 0.
```

Comment:
Should we swap the code (`for....`) for it's purpose? example: The I/O threads will busy-wait waiting for work ( writes or reads )?

Reply:
Replaced those code description